# TensorFlow 2.0 alpha - Convolutional Neural Networks
## Classify the MNIST Dataset

In [1]:
from __future__ import absolute_import, division, print_function

import tensorflow as tf
from tensorflow.keras import datasets, layers, models

/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [2]:
# Download and prepare MNIST dataset

(train_images, train_labels), (test_images, test_labels) = datasets.mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

# Normalize pixel values to be 0 - 1

train_images, test_images = train_images/255.0, test_images/255.0

### Create the Convolutional Base
#### Use a common pattern - stacking Conv2D and MaxPooling2D layers
#### the CNN receives Tensors input of shape (image_height, image_width, color_channels) - color images have 3 channels (R,G,B) - the MNIST images are Greyscale, with 1 channel
#### MNIST image input shape - (28, 28, 1)

In [4]:
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28,28,1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

In [5]:
# Display the Model Architecture

model.summary() 

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


#### Output from each layers is a 3D tensor - of shape (height, width, channels)

### Add Dense Layers on top
#### Dense layers take vectors of 1D - so the 3D tensors need to be Flattened - MNIST has 10 output classes

In [6]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten (Flatten)            (None, 576)               0         
_________________________________________________________________
dense (Dense)                (None, 64)               

#### the (3, 3, 64) outputs are Flattened into a vectors shaped (576) - then fed to Dense layers

## Compile and Train Model

In [8]:
model.compile(optimizer='adam',
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=5)

Epoch 1/5
60000/60000 [==============================] - 66s 1ms/sample - loss: 0.1446 - accuracy: 0.9559
Epoch 2/5
60000/60000 [==============================] - 81s 1ms/sample - loss: 0.0440 - accuracy: 0.9865
Epoch 3/5
60000/60000 [==============================] - 76s 1ms/sample - loss: 0.0336 - accuracy: 0.9894
Epoch 4/5
60000/60000 [==============================] - 77s 1ms/sample - loss: 0.0262 - accuracy: 0.9919
Epoch 5/5
60000/60000 [==============================] - 77s 1ms/sample - loss: 0.0201 - accuracy: 0.9935


## Evaluate the Model

In [10]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 5s 480us/sample - loss: 0.0346 - accuracy: 0.9892


In [11]:
print(test_acc)

0.9892


### This CNN has achieved a Test accuracy of over 98 %